kolibri/core/discovery/utils/network/search.py

```diff
@@ -0,0 +1,175 @@
+import atexit
+import json
+import logging
+import socket
+import time
+from contextlib import closing
+
+from zeroconf import get_all_addresses
+from zeroconf import NonUniqueNameException
+from zeroconf import ServiceInfo
+from zeroconf import USE_IP_OF_OUTGOING_INTERFACE
+from zeroconf import Zeroconf
+
+from kolibri.core.auth.models import Facility
+from kolibri.core.content.models import ChannelMetadata
+
+logger = logging.getLogger(__name__)
+
+SERVICE_TYPE = "_kolibri._http._tcp.local."
+LOCAL_DOMAIN = "kolibri.local"
+
+ZEROCONF_STATE = {"zeroconf": None, "listener": None, "service": None}
+
+
+def _id_from_name(name):
+    assert name.endswith(SERVICE_TYPE), (
+        "Invalid service name; must end with '%s'" % SERVICE_TYPE
+    )
+    return name.replace(SERVICE_TYPE, "").strip(".")
+
+
+def _is_port_open(host, port, timeout=1):
+    with closing(socket.socket(socket.AF_INET, socket.SOCK_STREAM)) as sock:
+        sock.settimeout(timeout)
+        return sock.connect_ex((host, port)) == 0
+
+
+class KolibriZeroconfService(object):
+
+    info = None
+
+    def __init__(self, id, port=8080, data={}):
+        self.id = id
+        self.port = port
+        self.data = {key: json.dumps(val) for (key, val) in data.items()}
+        atexit.register(self.cleanup)
+
+    def register(self):
+
+        if not ZEROCONF_STATE["zeroconf"]:
+            initialize_zeroconf_listener()
+
+        assert self.info is None, "Service is already registered!"
+
+        i = 1
+        id = self.id
```

Comment:
This assigns to a builtin, or is it just me?

Reply:
Seems so, yep! Perhaps I'll rename the local var to `unique_id`